### PR TITLE
SortedStorages refactoring

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Summary of Changes

Fixes #

### CODE REVIEW

File: runtime/src/sorted_storages.rs [hunk 1]
```diff
@@ -64,23 +64,20 @@ impl<'a> SortedStorages<'a> {
...
+        });
+        Self::new_with_slots(source.iter().zip(slots.into_iter()), None, None)
     }
 
     /// create `SortedStorages` from 'source' iterator.
     /// 'source' contains a SnapshotStorage and its associated slot
     /// 'source' does not have to be sorted in any way, but is assumed to not have duplicate slot #s
-    pub fn new_with_slots<'b>(
-        source: impl Iterator<Item = (&'a SnapshotStorage, &'b Slot)> + Clone,
+    pub fn new_with_slots(
```
Comment (jeffwashington) [inline runtime/src/sorted_storages.rs:79]: Main change is getting rid of `&'b` on Slot. Simplifies the api and was unnecessary.

File: runtime/src/sorted_storages.rs [hunk 2]
```diff
@@ -64,23 +64,20 @@ impl<'a> SortedStorages<'a> {
...
-            })
-            .collect::<Vec<_>>();
-        Self::new_with_slots(source.iter().zip(slots.iter()), None, None)
+        let slots = source.iter().map(|storages| {
+            let first = storages.first();
+            assert!(first.is_some(), "SnapshotStorage.is_empty()");
+            let storage = first.unwrap();
+            storage.slot() // this must be unique. Will be enforced in new_with_slots
+        });
+        Self::new_with_slots(source.iter().zip(slots.into_iter()), None, None)
```
Comment (yhchiang-sol) [inline runtime/src/sorted_storages.rs:73]: Neat!

File: runtime/src/sorted_storages.rs [hunk 3]
```diff
@@ -227,17 +224,18 @@ pub mod tests {
...
 
         pub fn new_for_tests(storages: &[&'a SnapshotStorage], slots: &[Slot]) -> Self {
-            SortedStorages::new_with_slots(storages.iter().cloned().zip(slots.iter()), None, None)
-        }
-
-        pub fn new_empty_for_tests() -> Self {
-            SortedStorages::new(&[])
+            assert_eq!(storages.len(), slots.len());
+            SortedStorages::new_with_slots(
+                storages.iter().cloned().zip(slots.iter().cloned()),
```
Comment (yhchiang-sol) [inline runtime/src/sorted_storages.rs:229]: I guess this clone might be what it takes to avoid a lifetime specifier, but is there any way to avoid it?

**Review by yhchiang-sol — APPROVED** [review]
Looks good to me.  Only one question about the additional clone.

File: runtime/src/sorted_storages.rs (hunk 3 above)
Comment (jeffwashington) [reply runtime/src/sorted_storages.rs:229]: this doesn't avoid the lifetime specifier. It is just there to get rid of references and make the types work. The clone is only in test code. Perhaps there is a rust way to get rid of the clone. I tried a few and the compiler and clippy both had opinions.

**Review by HaoranYi — COMMENTED** [review]
looks good. Thanks!

**Pull request merged**